### ISSUE TITLE
Do these Metrics really need to be scaled?

### ISSUE THREAD

**Issue by fabianlim** [verl-project/verl#3129]
I noticed there is some scaling with `loss_scale_factor` in the `dp_actor`:
- [pg_loss](https://github.com/volcengine/verl/blob/afd759789bd8b80b692361ca971758a0f34d75da/verl/workers/actor/dp_actor.py#L477)
- [kl_loss](https://github.com/volcengine/verl/blob/afd759789bd8b80b692361ca971758a0f34d75da/verl/workers/actor/dp_actor.py#L465)

```python 
     "actor/pg_loss": pg_loss.detach().item() * loss_scale_factor,
```

However I am wondering if these scalings are really needed, since the [reduction](https://github.com/volcengine/verl/blob/afd759789bd8b80b692361ca971758a0f34d75da/verl/utils/metric/utils.py#L23) over the metrics is by `mean` and not `sum`?

**Comment by vermouth1992** [comment]
The loss is scaled so that they are parallelism ignorant. Otherwise, you will find that these losses magnitude is different if you use different sp_size, which should not be the case

**Comment by fabianlim** [comment]
> The loss is scaled so that they are parallelism ignorant. Otherwise, you will find that these losses magnitude is different if you use different sp_size, which should not be the case

@vermouth1992 yes I agree the loss should be scaled for the backward, but this was not my point. My point is only that for **reporting**, we shouldnt need to scale it becasue there is already a reduction outside by **mean**. 

The key point you need to scale it for the backward is because multiple forward-backward calls will reduce by **sum**, but that is not the case for reporting.

**Comment by vermouth1992** [comment]
> > The loss is scaled so that they are parallelism ignorant. Otherwise, you will find that these losses magnitude is different if you use different sp_size, which should not be the case
> 
> [@vermouth1992](https://github.com/vermouth1992) yes I agree the loss should be scaled for the backward, but this was not my point. My point is only that for **reporting**, we shouldnt need to scale it becasue there is already a reduction outside by **mean**.
> 
> The key point you need to scale it for the backward is because multiple forward-backward calls will reduce by **sum**, but that is not the case for reporting.

I see what you mean. Maybe I was wrong. Could you confirm that those values will not change its magnitude when the ulysses sp size changes? Thanks!

**Comment by fabianlim** [comment]
@vermouth1992 its kind of roughly the same

```

> metrics['actor/pg_loss']
[-0.3054201006889343, 0.0, -0.2829209864139557, -0.16481846570968628, -0.15159031748771667]
> self.ulysses_sequence_parallel_size
1


> metrics['actor/pg_loss']
[1.2069861888885498, -0.24875977635383606, -0.2340698093175888, -0.1880173534154892, 0.0, -0.07240534573793411, 0.0, 0.0, -0.1661381721496582]
> self.ulysses_sequence_parallel_size
2

```

**Comment by hiyouga** [comment]
I think the current implementation has some bugs, we'll fix it soon

**Comment by Siyuexi** [comment]
Agree. Current metrics looks confusing.

**Comment by Ber666** [comment]
Agree

**Comment by xuuHuang** [comment]
Agree. When using different ppo_micro_batch_size_per_gpu, the magnitude is different.

**Comment by Kirrito-k423** [comment]
Thanks for reporting this. I rechecked the current code and related PR history, and this looks resolved.

English:
- This issue is about `actor/pg_loss` / `actor/kl_loss` metric scaling under micro-batches / gradient accumulation.
- PR https://github.com/verl-project/verl/pull/4711 was merged on 2025-12-30 and specifically fixed the extra scaling in logged KL/PG losses. Its validation showed that reducing `ppo_micro_batch_size_per_gpu` from 8 to 2 changed the logged loss by 4x before the fix, while the fixed branch no longer changed the metric magnitude.
- On current `main`, the old `verl/workers/actor/dp_actor.py` path is gone. The current engine path records `actor/pg_loss` as a `Metric` with SUM/MEAN aggregation depending on global-batch normalization, and `reduce_metrics` respects `Metric.aggregate()` instead of blindly averaging already-scaled plain values.

中文：
- 这个 issue 讨论的是 `actor/pg_loss` / `actor/kl_loss` 在 micro-batch / 梯度累积场景下的指标缩放问题。
- https://github.com/verl-project/verl/pull/4711 已在 2025-12-30 合并，专门修复了 KL/PG logged losses 的额外缩放。该 PR 的验证显示，修复前 `ppo_micro_batch_size_per_gpu` 从 8 降到 2 会让 logged loss 变化 4 倍，修复分支中该指标量级不再随这个配置变化。
- 当前 `main` 已经没有旧的 `verl/workers/actor/dp_actor.py` 路径；新 engine 路径用带 SUM/MEAN 语义的 `Metric` 记录 `actor/pg_loss`，`reduce_metrics` 会按 `Metric.aggregate()` 聚合，不再对已经缩放的普通值列表做盲目均值。

Based on those public code and PR references, I think this can be closed as fixed.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
